##SQL+Python购物篮关联分析

In [46]:
#本实验所使用的包
from efficient_apriori import apriori
import sqlalchemy as sql
import pandas as pd

In [47]:
#连接到数据库
engine = sql.create_engine('mysql+mysqlconnector://root:password@localhost:3306/zhiwen')

In [48]:
#获取全部的数据
sql = 'SELECT * FROM bread_basket'
data = pd.read_sql(query, engine)

In [49]:
#对数据进行基本分析
data.head()

Date      Time Transaction           Item
0  2016/10/30   9:58:11           1           NONE
1  2016/10/30  10:05:34           2   Scandinavian
2  2016/10/30  10:05:34           2   Scandinavian
3  2016/10/30  10:07:57           3  Hot chocolate
4  2016/10/30  10:07:57           3            Jam

In [50]:
#调整Item项的格式
data['Item']=data['Item'].str.lower()

In [51]:
data.head()

Date      Time Transaction           Item
0  2016/10/30   9:58:11           1           none
1  2016/10/30  10:05:34           2   scandinavian
2  2016/10/30  10:05:34           2   scandinavian
3  2016/10/30  10:07:57           3  hot chocolate
4  2016/10/30  10:07:57           3            jam

In [52]:
#删除none数据
data = data.drop(data[data['Item']== 'none'].index)

In [53]:
data.head()

Date      Time Transaction           Item
1  2016/10/30  10:05:34           2   scandinavian
2  2016/10/30  10:05:34           2   scandinavian
3  2016/10/30  10:07:57           3  hot chocolate
4  2016/10/30  10:07:57           3            jam
5  2016/10/30  10:07:57           3        cookies

In [54]:
data.tail()

Date      Time Transaction       Item
21288  2017/4/9  14:32:58        9682     coffee
21289  2017/4/9  14:32:58        9682        tea
21290  2017/4/9  14:57:06        9683     coffee
21291  2017/4/9  14:57:06        9683     pastry
21292  2017/4/9  15:04:24        9684  smoothies

In [55]:
# 得到一维数组orders_series，并且将Transaction作为index, value为Item取值
orders_series = data.set_index('Transaction')['Item']

In [56]:
# 将数据集进行格式转换，转化成关联分析需要的格式
transactions = []
temp_index = 0
for i, v in orders_series.items():
	if i != temp_index:
		temp_set = set()
		temp_index = i
		temp_set.add(v)
		transactions.append(temp_set)
	else:
		temp_set.add(v)

In [57]:
# 挖掘频繁项集和频繁规则
itemsets, rules = apriori(transactions, min_support=0.02,  min_confidence=0.5)
print('频繁项集：', itemsets)
print('关联规则：', rules)

频繁项集： {1: {('alfajores',): 344, ('bread',): 3096, ('brownie',): 379, ('cake',): 983, ('coffee',): 4528, ('cookies',): 515, ('farm house',): 371, ('hot chocolate',): 552, ('juice',): 365, ('medialuna',): 585, ('muffin',): 364, ('pastry',): 815, ('sandwich',): 680, ('scandinavian',): 275, ('scone',): 327, ('soup',): 326, ('tea',): 1350, ('toast',): 318, ('truffles',): 192}, 2: {('bread', 'cake'): 221, ('bread', 'coffee'): 852, ('bread', 'pastry'): 276, ('bread', 'tea'): 266, ('cake', 'coffee'): 518, ('cake', 'tea'): 225, ('coffee', 'cookies'): 267, ('coffee', 'hot chocolate'): 280, ('coffee', 'juice'): 195, ('coffee', 'medialuna'): 333, ('coffee', 'pastry'): 450, ('coffee', 'sandwich'): 362, ('coffee', 'tea'): 472, ('coffee', 'toast'): 224}}
关联规则： [{cake} -> {coffee}, {cookies} -> {coffee}, {hot chocolate} -> {coffee}, {juice} -> {coffee}, {medialuna} -> {coffee}, {pastry} -> {coffee}, {sandwich} -> {coffee}, {toast} -> {coffee}]


##实验结果：频繁项集： {1: {('alfajores',): 344, ('bread',): 3096, ('brownie',): 379, ('cake',): 983, ('coffee',): 4528, ('cookies',): 515, ('farm house',): 371, ('hot chocolate',): 552, ('juice',): 365, ('medialuna',): 585, ('muffin',): 364, ('pastry',): 815, ('sandwich',): 680, ('scandinavian',): 275, ('scone',): 327, ('soup',): 326, ('tea',): 1350, ('toast',): 318, ('truffles',): 192}, 2: {('bread', 'cake'): 221, ('bread', 'coffee'): 852, ('bread', 'pastry'): 276, ('bread', 'tea'): 266, ('cake', 'coffee'): 518, ('cake', 'tea'): 225, ('coffee', 'cookies'): 267, ('coffee', 'hot chocolate'): 280, ('coffee', 'juice'): 195, ('coffee', 'medialuna'): 333, ('coffee', 'pastry'): 450, ('coffee', 'sandwich'): 362, ('coffee', 'tea'): 472, ('coffee', 'toast'): 224}}
关联规则： [{cake} -> {coffee}, {cookies} -> {coffee}, {hot chocolate} -> {coffee}, {juice} -> {coffee}, {medialuna} -> {coffee}, {pastry} -> {coffee}, {sandwich} -> {coffee}, {toast} -> {coffee}]